pkg/rbac/parser.go

```diff
@@ -63,83 +61,93 @@ type Rule struct {
 	Namespace string `marker:",optional"`
 }
 
-// ruleKey represents the resources and non-resources a Rule applies.
-type ruleKey struct {
-	Groups        string
-	Resources     string
-	ResourceNames string
-	URLs          string
-}
+func newSet(strings []string) sets.String {
+	if strings == nil { // preserve nils for back-compat with existing code
+		return nil
+	}
 
-func (key ruleKey) String() string {
-	return fmt.Sprintf("%s + %s + %s + %s", key.Groups, key.Resources, key.ResourceNames, key.URLs)
+	return sets.NewString(strings...)
 }
 
-// ruleKeys implements sort.Interface
-type ruleKeys []ruleKey
-
-func (keys ruleKeys) Len() int           { return len(keys) }
-func (keys ruleKeys) Swap(i, j int)      { keys[i], keys[j] = keys[j], keys[i] }
-func (keys ruleKeys) Less(i, j int) bool { return keys[i].String() < keys[j].String() }
-
-// key normalizes the Rule and returns a ruleKey object.
-func (r *Rule) key() ruleKey {
-	r.normalize()
-	return ruleKey{
-		Groups:        strings.Join(r.Groups, "&"),
-		Resources:     strings.Join(r.Resources, "&"),
-		ResourceNames: strings.Join(r.ResourceNames, "&"),
-		URLs:          strings.Join(r.URLs, "&"),
+func (r *Rule) Normalize() *NormalizedRule {
+	result := &NormalizedRule{
+		Groups:        newSet(r.Groups),
+		Resources:     newSet(r.Resources),
+		ResourceNames: newSet(r.ResourceNames),
+		Verbs:         newSet(r.Verbs),
+		URLs:          newSet(r.URLs),
+		Namespace:     r.Namespace,
+	}
+
+	// fix the group names, since letting people type "core" is nice
+	if _, ok := result.Groups["core"]; ok {
+		delete(result.Groups, "core")
+		result.Groups[""] = struct{}{}
 	}
+
+	result.GenerateComparisonKey()
+
+	return result
 }
 
-// addVerbs adds new verbs into a Rule.
-// The duplicates in `r.Verbs` will be removed, and then `r.Verbs` will be sorted.
-func (r *Rule) addVerbs(verbs []string) {
-	r.Verbs = removeDupAndSort(append(r.Verbs, verbs...))
+type NormalizedRule struct {
+	// if two different rules have the same comparison key then they can have their verbs merged
+	// this key should comprise all the fields below except Verbs and Namespace (since rules
+	// are partitioned by namespace before merging)
+	ComparisonKey string
```

Comment:
Sorry, I don't understand why this is needed. If one rule is a superset of another rule, the subet rule can be removed. Why is any additional merging needed?

Reply:
This is to handle verb list merging (a part of the existing codebase as well); if two rules match exactly the same things then they can be unified and their verb lists merged.